### PR TITLE
External textures wait API

### PR DESCRIPTION
**Pull request by markschlosseratbentley**
This PR adds API to RenderSystem which returns a promise which when resolved indicates that all pending external textures have finished loading.

### CODE REVIEW

**Review by pmconne — COMMENTED** [review]
I was hoping for an async Viewport method that would resolve after everything required to draw the viewport's scene (tile trees, tiles, textures, etc) are loaded.

File: core/frontend/src/render/RenderSystem.ts [hunk 1]
```diff
@@ -391,6 +391,9 @@ export abstract class RenderSystem implements IDisposable {
    */
   public abstract createBatch(graphic: RenderGraphic, features: PackedFeatureTable, range: ElementAlignedBox3d, options?: BatchOptions): RenderGraphic;
 
+  /** Return a Promise which when resolved indicates that all pending external textures have finished loading from the backend. */
+  public abstract waitForAllExternalTextures(): Promise<void>;
```
Comment (pmconne) [inline core/frontend/src/render/RenderSystem.ts:395]: Make this non-abstract and make it return `Promise.resolve` so only System has to override it.

File: core/frontend/src/render/webgl/System.ts [hunk 2]
```diff
@@ -404,6 +404,19 @@ export class System extends RenderSystem implements RenderSystemDebugControl, Re
     return this.techniques.idleCompileNextShader();
   }
 
+  /** Return a Promise which when resolved indicates that all pending external textures have finished loading from the backend. */
+  public async waitForAllExternalTextures(): Promise<void> {
+    const extTexLoader = ExternalTextureLoader.instance;
+    if (extTexLoader.numActiveRequests < 1 && extTexLoader.numPendingRequests < 1)
+      return Promise.resolve();
+    const promise = new Promise<void>((resolve: any) => {
+      extTexLoader.onTexturesLoaded.addListener(() => {
```
Comment (pmconne) [inline core/frontend/src/render/webgl/System.ts:413]: `/s/addListener/addOnce`, no?

**Review by pmconne — CHANGES_REQUESTED** [review]
Useful tests:
- What happens when no external texture requests exist.
- What happens when 1 or more requests exist.
- What happens if a request throws.

**Comment by markschlosseratbentley** [comment]
> I was hoping for an async Viewport method that would resolve after everything required to draw the viewport's scene (tile trees, tiles, textures, etc) are loaded.

@pmconne That sounds useful.  Think that should be this PR, or a future one?

**Comment by markschlosseratbentley** [comment]
> I was hoping for an async Viewport method that would resolve after everything required to draw the viewport's scene (tile trees, tiles, textures, etc) are loaded.

@pmconne - I see TileAdmin has onTileLoad, onTileTreeLoad, onTileChildrenLoad - do you think having the new Viewport function use these in addition to the external texture wait function would be a good approach?

**Comment by pmconne** [comment]
>  I see TileAdmin has onTileLoad, onTileTreeLoad, onTileChildrenLoad - do you think having the new Viewport function use these in addition to the external texture wait function would be a good approach?

No, those events are dispatched when individual tiles/trees load.

Example use case is: app wants to capture an image of the viewport. For example, to use as a view thumbnail, to serve as input to ImageTests, etc.
Instead of having to copy all the complicated looping and re-rendering done by display-performance-test-app's TestRunner, app should be able to await a single Viewport method that will handle all of that internally.

**Comment by markschlosseratbentley** [comment]
> > I see TileAdmin has onTileLoad, onTileTreeLoad, onTileChildrenLoad - do you think having the new Viewport function use these in addition to the external texture wait function would be a good approach?
> 
> No, those events are dispatched when individual tiles/trees load.
> 
> Example use case is: app wants to capture an image of the viewport. For example, to use as a view thumbnail, to serve as input to ImageTests, etc.
> Instead of having to copy all the complicated looping and re-rendering done by display-performance-test-app's TestRunner, app should be able to await a single Viewport method that will handle all of that internally.

@pmconne - Yes, that kind of method sounds helpful.  I am wondering the best way to accomplish it.  Should Tile have events that they pass upward to parents, etc.?  Or could we use those TileAdmin events to determine when we should check if everything is done loading?  Doing a loop with various timers like is done in TestRunner, etc. seems a bit clunky....

**Comment by pmconne** [comment]
@markschlosseratbentley why not finish this PR and assign a backlog item to me for the Viewport API?

**Comment by markschlosseratbentley** [comment]
> @markschlosseratbentley why not finish this PR and assign a backlog item to me for the Viewport API?

Sure.  Thanks.

File: core/frontend/src/render/RenderSystem.ts (hunk 1 above)
Comment (markschlosseratbentley) [reply core/frontend/src/render/RenderSystem.ts]: Done - that makes it much simpler.

File: core/frontend/src/render/webgl/System.ts (hunk 2 above)
Comment (markschlosseratbentley) [reply core/frontend/src/render/webgl/System.ts]: Yeah, you are correct. Done.

**Comment by markschlosseratbentley** [comment]
> Useful tests:
> 
> * What happens when no external texture requests exist.
> * What happens when 1 or more requests exist.
> * What happens if a request throws.

I added another wait in the ExternalTextures.test.ts which verifies nothing ill happens when waiting when no external texture requests exist.

The same test already tests what happens when 1 or more requests exist.

The same test also verifies that the wait works properly when querying bad texture names.

**Review by pmconne — APPROVED** [review]

**Pull request merged**